tailfs/tailfsimpl/dirfs/dirfs_test.go

```diff
@@ -0,0 +1,346 @@
+			label: "attempt to create static root folder",
+			name:  "/domain",
+		},
+		{
+			label: "attempt to create remote",
+			name:  "/domain/remote1",
+		},
+		{
+			label: "attempt to create non-existent remote",
+			name:  "/domain/remote3",
+			err:   os.ErrPermission,
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			err := fs.Mkdir(ctx, test.name, test.perm)
+			if test.err != nil {
+				if !errors.Is(err, test.err) {
+					t.Errorf("got %v, want %v", err, test.err)
+				}
+			} else if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+		})
+	}
+}
+
+func TestRemoveAll(t *testing.T) {
+	fs, _, _, _ := createFileSystem(t)
+
+	tests := []struct {
+		label string
+		name  string
+		err   error
+	}{
+		{
+			label: "attempt to remove root folder",
+			name:  "/",
+			err:   os.ErrPermission,
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			err := fs.RemoveAll(ctx, test.name)
+			if !errors.Is(err, test.err) {
+				t.Errorf("got %v, want %v", err, test.err)
+			}
+		})
+	}
+}
+
+func TestRename(t *testing.T) {
+	fs, _, _, _ := createFileSystem(t)
+
+	tests := []struct {
+		label   string
+		oldName string
+		newName string
+		err     error
+	}{
+		{
+			label:   "attempt to move root folder",
+			oldName: "/",
+			newName: "/domain/remote2/copy.txt",
+			err:     os.ErrPermission,
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		t.Run(test.label, func(t *testing.T) {
+			err := fs.Rename(ctx, test.oldName, test.newName)
+			if !errors.Is(err, test.err) {
+				t.Errorf("got %v, want: %v", err, test.err)
+			}
+		})
+	}
+}
+
+func createFileSystem(t *testing.T) (webdav.FileSystem, string, string, *tstest.Clock) {
+	s1, dir1 := startRemote(t)
+	s2, dir2 := startRemote(t)
+
+	// Make some files, use perms 0666 as lowest common denominator that works
+	// on both UNIX and Windows.
+	err := os.WriteFile(filepath.Join(dir1, "file1.txt"), []byte("12345"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.WriteFile(filepath.Join(dir2, "file2.txt"), []byte("54321"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// make some directories
+	err = os.Mkdir(filepath.Join(dir1, "dir1"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.Mkdir(filepath.Join(dir2, "dir2"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	clock := tstest.NewClock(tstest.ClockOpts{Start: time.Now()})
+	fs := &FS{
+		Clock:      clock,
+		StaticRoot: "domain",
+		Children: []*Child{
+			{Name: "remote1"},
+			{Name: "remote2"},
+			{Name: "remote4"},
+		},
+	}
+
+	t.Cleanup(func() {
+		defer s1.Close()
+		defer os.RemoveAll(dir1)
+		defer s2.Close()
+		defer os.RemoveAll(dir2)
+	})
+
+	return fs, dir1, dir2, clock
+}
+
+func startRemote(t *testing.T) (*httptest.Server, string) {
+	dir := t.TempDir()
+
+	h := &webdav.Handler{
+		FileSystem: webdav.Dir(dir),
+		LockSystem: webdav.NewMemLS(),
+	}
+
```

Comment:
I'd suggest also:

```go
s := httptest.NewServer(h)
t.Cleanup(s.Close)
return s, dir
```

so it will get auto-cleaned-up when the test unwinds.